lib/validator/task/mod.rs

```diff
@@ -1155,15 +1155,446 @@ where
+
+        // Insert new block
+        self.blocks.insert(block_hash, block);
+        self.insertion_order.push_back(block_hash);
+        tracing::trace!(
+            "Cached block `{}` (cache size: {})",
+            block_hash,
+            self.blocks.len()
+        );
+    }
+
+    fn remove(&mut self, block_hash: &BlockHash) -> Option<Block> {
+        if let Some(block) = self.blocks.remove(block_hash) {
+            // Remove from insertion order
+            if let Some(pos) = self.insertion_order.iter().position(|h| h == block_hash) {
+                self.insertion_order.remove(pos);
+            }
+            tracing::trace!("Cache hit for block `{}`", block_hash,);
+            Some(block)
+        } else {
+            None
+        }
+    }
+
+    fn len(&self) -> usize {
+        self.blocks.len()
+    }
+}
+
+/// Configuration constants for block file parsing with caching
+const MAX_BLOCK_CACHE_SIZE: usize = 5000;
+const MAX_ITERATIONS_WITHOUT_MATCH: usize = 5000;
+const BLOCKS_DIR_CONNECT_BATCH_SIZE: usize = 2000;
+
+/// Check cache for subsequent expected blocks and process them if found.
+/// This function iteratively looks for the next expected block in the cache
+/// and processes it until no more consecutive blocks are found.
+fn process_cached_blocks(
+    dbs: &Dbs,
+    event_tx: &Sender<Event>,
+    block_cache: &mut BlockCache,
+    missing_blocks: &mut Vec<BlockHash>,
+    pending_blocks: &mut Vec<Block>,
+    total_handled_blocks: &mut usize,
+) -> Result<(), error::Sync> {
+    while let Some(&expected_block_hash) = missing_blocks.last() {
+        if let Some(cached_block) = block_cache.remove(&expected_block_hash) {
+            // Found next expected block in cache
+            pending_blocks.push(cached_block);
+            missing_blocks.pop();
+
+            // Check if we should process batch
+            if pending_blocks.len() >= BLOCKS_DIR_CONNECT_BATCH_SIZE || missing_blocks.is_empty() {
+                handle_block_batch(dbs, pending_blocks, event_tx)?;
+                *total_handled_blocks += pending_blocks.len();
+                pending_blocks.clear();
+            }
+        } else {
+            // Next expected block not in cache, break out
+            break;
+        }
+    }
+
+    Ok(())
+}
+
+/// Sync blocks by reading from the raw block files. The idea is to mutate the missing
+/// blocks vector, such that once this function returns the JSON-RPC sync logic can
+/// finish off the job. Syncing from the block directory is not expected to work all
+/// until the end due to Core having blocks it hasn't flushed to disk.
+///
+/// Handles non-sequential block layout by temporarily caching unexpected blocks
+/// and checking the cache when expected blocks are found.
+#[tracing::instrument(skip_all)]
+async fn sync_from_blocks_dir<Signal>(
+    dbs: &Dbs,
+    event_tx: &Sender<Event>,
+    missing_blocks: &mut Vec<BlockHash>,
+    main_blocks_dir: PathBuf,
+    shutdown_signal: Signal,
```

Comment:
It might be good to use [tokio_util::sync::CancellationToken](https://docs.rs/tokio-util/latest/tokio_util/sync/struct.CancellationToken.html) as a more explicit shutdown signal.
IIUC, This function is only `async` so that we can poll the shutdown signal, whereas `CancellationToken` can be checked synchronously. Maybe we should do that in a different PR.

For now, it would be good to poll `shutdown_signal` manually, which would avoid having to make this function `async`.